web/api/v1/api.go

```diff
@@ -1793,16 +1793,20 @@ func (api *API) respond(w http.ResponseWriter, req *http.Request, data interface
 }
 
 func (api *API) negotiateCodec(req *http.Request, resp *Response) (Codec, error) {
-	for _, clause := range goautoneg.ParseAccept(req.Header.Get("Accept")) {
-		for _, codec := range api.codecs {
-			if codec.ContentType().Satisfies(clause) && codec.CanEncode(resp) {
+	for _, codec := range api.codecs {
+		if !codec.CanEncode(req, resp) {
+			continue
+		}
+
+		for _, clause := range goautoneg.ParseAccept(req.Header.Get("Accept")) {
+			if codec.ContentType().Satisfies(clause) {
```

Comment:
Why doesn't this disappear inside `CanEncode` ?

Reply:
The Codec interface is build around mime types (through `ContentType() MIMEType` here and in `w.Header().Set("Content-Type", codec.ContentType().String())`), if we're not going to use/enforce that, maybe we should also adjust the interface.

My suggestion was to change `negotiateCodec` so that it asks first if the Codec canEncode, but also make sure that the coded still satisfies the mimetype.

Reply:
I think it is cleaner if mime type matching is moved to `CanEncode` like Bryan suggested.

It also makes sure that the default codec will not be chosen if the request does not have an accept header for it. Currently the default codec can be chosen even if the mime type does not match (if there was no other valid codec). 

You make a good point though since the implementation of mime type checking in `CanEncode` is not enforced in the interface. Is that something we should worry about or would adding a `MimeTypeCheck` helper function and documenting that the codec implementations of `CanEncode` have the responsibility to make sure mime type matches would be enough.